File: deno/src/base/org/org.service.ts
import type {
  Mutation,
} from "/gen/types.ts"

import {
  getAuthModel,
} from "/lib/auth/auth.dao.ts";

import {
  findByIdUsr,
} from "/gen/base/usr/usr.dao.ts";

import {
  createToken,
} from "/lib/auth/auth.dao.ts";

import {
  ns,
} from "/src/base/i18n/i18n.ts";

export async function orgLoginSelect(
  org_id?: OrgId,
): Promise<Mutation["orgLoginSelect"]> {
  
  const authModel = await getAuthModel();
  if (!authModel) {
    throw await ns("用户未登录");
  }
  if (!org_id && !authModel.org_id) {
    return "";
  }
  if (authModel.org_id === org_id) {
    return "";
  }
  if (org_id) {
    const usr_model = await findByIdUsr(authModel.id);
    const org_ids = usr_model?.org_ids || [ ];
    if (!org_ids.includes(org_id)) {
      throw await ns("无权限切换到该组织");
    }
  }
  authModel.org_id = org_id;
  // authModel.exp = undefined;
  const { authorization } = await createToken(authModel);
  return authorization;
}

File: rust/src/src/base/org/org_service.rs
use anyhow::Result;

use crate::common::auth::auth_dao;
use crate::common::context::{
  Ctx,
  get_auth_model,
  get_auth_org_id,
};
use crate::common::id::ID;

use crate::gen::base::usr::usr_dao;

pub async fn org_login_select(
  ctx: &mut Ctx,
  org_id: ID,
) -> Result<String> {
  let org_id2 = get_auth_org_id();
  if let Some(org_id2) = org_id2 {
    if org_id == org_id2 {
      return Ok("".to_owned());
    }
  }
  let mut auth_model = get_auth_model()
    .ok_or_else(|| 
      anyhow::anyhow!("auth_model.is_none()")
    )?;
  let usr_model = usr_dao::find_by_id(
    auth_model.id.clone(),
    None,
  ).await?;
  let org_ids: Vec<ID> = {
    if let Some(usr_model) = usr_model {
      usr_model.org_ids
    } else {
      vec![]
    }
  };
  if !org_ids.contains(&org_id) {
    return Err(anyhow::anyhow!("org_id: {org_id} dose not exit in login usr"));
  }
  auth_model.org_id = org_id.into();
  let token = auth_dao::get_token_by_auth_model(&auth_model)?;
  ctx.set_auth_model(auth_model);
  Ok(token)
}
